pkg/controller/node_availability.go

```diff
@@ -0,0 +1,155 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+	"sync/atomic"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/klog/v2"
+	azClientSet "sigs.k8s.io/azuredisk-csi-driver/pkg/apis/client/clientset/versioned"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+type ReconcileNodeAvailability struct {
+	client                client.Client
+	controllerSharedState *SharedState
+	azVolumeClient        azClientSet.Interface
+	kubeClient            kubernetes.Interface
+}
+
+var _ reconcile.Reconciler = &ReconcileNodeAvailability{}
+
+func (r *ReconcileNodeAvailability) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+
+	n := &corev1.Node{}
+	err := r.client.Get(ctx, request.NamespacedName, n)
+
+	if errors.IsNotFound(err) {
+		return reconcile.Result{}, nil
+	}
+
+	if err == nil {
+		if !n.Spec.Unschedulable {
+			for _, taint := range n.Spec.Taints {
```

Comment:
Taints don't prevent every pod from being scheduled to this node, so we shouldn't exclude it here based on the presence of a no-schedule taint.

Reply:
Thats a good point, I will update this to just check for the node.kubernetes.io/unschedulable taint, as that should still allow us to requeue the request until the cache with this taint is updated to match what is reflected in n.Spec.Unschedulable

Reply:
Good to know about the `node.kubernetes.io/unschedulable` taint. You'll need to update the update predicate on line 121 to return `true` when taints change then. (The `Unschedulable` property is updated by the user; presumably, the taint is added or removed by the Node reconciler in response. This means we might see the change in Unschedulable before the taint has been added/removed.)

Reply:
As another thought, should we also be using the node_availability controller to try to add replicas when other taints are updated other than the basic node.kubernetes.io/unschedulable node scheduling? For example, when any taint gets updated on the node, the node_availability controller could try to add replica attachments from the priority queue to see if any are newly tolerated

Reply:
I don't believe it very likely that taints will be changed often, and when they do, it mostly to support new workloads that have yet to be scheduled. I don't believe we need to consider that scenario in this PR nor in the future.